### DISCUSSION TITLE
Merge assessment-store into pre-award-stores

### DISCUSSION THREAD

**Discussion by srh-sloan** [communitiesuk/funding-service-requests-for-comments#8] [General]
```
Title: Merge assessment-store into pre-award-stores
Owner: @srh-sloan @sfount 
Collaborator(s): @samuelhwilliams @MarcUsher 
Created on: 20/11/2024
Status: Draft
Finalised on: 
```

## Overview

The `fund-store` and `application-store` have been merged as part of an [earlier RFC](https://github.com/communitiesuk/funding-service-requests-for-comments/discussions/4).

With the larger goal of reducing the number of databases and stores we are also looking to merge `assessment-store` into this new combined stores service. This is also the next step in completing the work for this [overarching RFC](https://github.com/communitiesuk/funding-service-requests-for-comments/discussions/3) to consolidate the pre-award services into a monolith

## What is the current state?

The pre-award services have previously used 4 separate stores and databases. The work to merge them into one store is part way through, as follows:

| Name | Future Location | Status |
|--------|--------|--------|
| fund-store | pre-award-stores | Code complete, deployments in progress |
| application-store | pre-award-stores | Ready to start, once fund-store complete |
| assessment-store | pre-award-stores | Not started (this RFC) |
| account-store | TBD | Not started |

Most of these stores include some form of fund configuration, with some technical infrastructure existing between the stores themselves, and all four are talked to by both the pre-award front ends (assessment and application).

## Why should we change?

Alongside all of the challenges described in the referenced RFCs, separating the application and assessment creates two copies of the same application data, currently separated by a queue. The addition of this queue to separate application and assessment data increases complexity in the pre-award estate and adds to the number of moving parts involved in development work.

Having separate data and configuration for applications and assessments also encourages specific/ fixed mappings of data for the apply and assess process which makes the amount of configuration and work needed for onboarding challenging. 

Separate data and spread out configuration can also lead to hard to track down edge cases and loses any transactional benefits of working in a standard relational database.

Merging the application store and assessment store will in time (future RFC) allow us to use a single copy of the application data, reduce the amount of configuration needed for a new fund and make the development and testing process easier and smoother.

### Phase 1

1. Lift and shift the existing `assessment-store` into `pre-award-stores`. This will follow a very similar process to combining the `fund-store` and `application-store` in that code will be kept in 2 largely separate modules initially. 
    i. The application-store and assessment-store will continue to communicate via a queue, even though they are co-located.  
    ii. The old assessment-store repo can be archived
    iii. Migrate the data from `assessment-store` into matching tables in `pre-award-stores`.
    iv. Update the local docker-runner setup to use pre-award-stores instead of assessment-store
5. Refine the interface for moving an application from "apply" to "assess". Instead of interfacing with SQS this will directly copy data from the "application" table to the "assessment" table in the pre-award store.
    i. At this stage the queue is no longer in use

### Phase 2
- Remove the old `assessment-store` app and database from AWS
- Remove the application import queue from AWS
- Remove the old queueing code from the `application-store` module in `pre-award-stores`
- Remove the old assessment-store service from the docker runner

### Phase 3
Establish how we can take advantage of the new merged assess and apply data to further simplify configuration and onboarding. This will be the subject of a future spike and RFC. 

## What alternatives are we discarding?

 ❌ Keep things as they are - everyone has agreed that the existing architecture is too complicated and too interconnected, not giving us the benefits of true microservice architecture and being over-engineered for the scale of the services
 ❌ Consolidating the services but not merging databases - while this might simplify deployments it wouldn't address the service needs around consistent data and a significantly simpler developer experience

## Who will be affected?

At this stage it will be just the funding service developers that will be affected - they will have one less application to configure and run locally, and one less deployment. 

Users should notice no tangible difference.

## Who will benefit?

Initially, just the funding service developers. But this change is a step on the road to simplifying the overall data model and configuration which will lead to shorter onboarding times, less technical complexity for the funding service teams and a more maintainable service.

## What are the key risks to manage or mitigate?

- This work depends on the previous RFCs to merge fund-store and application-store. This work is progressing well but it should be monitored to ensure it does not delay this RFC
- Migrating the data from one database to another - this process will be tested in 2 ways to mitigate risks. The overall approach will have been tested during the fund-store and application-store migrations, and the specific migration of assessment data will be tested in non-prod environments before we migrate production data. Database backups exist (and will be checked) before we do a full migration.
- As this work will involved a code freeze on the `assessment-store`, this will affect other funding service teams. Coordination will be required between platforms and live services to plan a time for this freeze to minimise disruption and conflicts.

**Comment by gidsg** [comment]
> Add a feature flag to the application-store (in pre-award-stores) to either send new applications to the existing queue infrastructure (picked up by existing assessment-store service), or use the new interface to copy data directly into the new assessment tables in pre-award-stores.

Could we even do both in parallel to validate the results, i.e. closer to a 'dark launch' pattern? Or would they be writing to the same place?

**Reply by srh-sloan** [reply]
Changed this approach now after talking with Sam

**Comment by samuelhwilliams** [comment]
> This is the next step in completing the work for this [overarching RFC](https://github.com/communitiesuk/funding-service-requests-for-comments/discussions/3) to consolidate all the pre-award stores into one service (pre-award-stores)

The RFC quoted is about consolidating all of pre-award into a monolith, rather than just the stores, I think?

> Since the queueing infrastructure for moving applications from apply to assess will no longer be required, any code relating to this will not be part of the initial 'lift & shift' phase, it will be left behind in assessment-store.

I think this bit is still giving me pause, and it'd be useful to chat some of this through as it may be easier to have a back and forth face-to-face (but also maybe some of our discussion can then be written up here for future reference).

Phase 1 step 1i suggests that we immediately drop the queueing behaviour - ie don't bring any of that into pre-award-stores. But the application-store lifted+shifted into pre-award-stores will still be writing to the queue. Phase 2 step 1 then talks about adding a feature flag to the application_store module in pre-award-stores to toggle between sending to a queue (to be picked up by old assessment-store) and writing to the DB directly. Except by this point, I think the old assessment-store service will have been scaled down to zero and turned off? So nothing would read that queue.

I can't tell if these two steps are directly conflicting, or if the explanation is maybe unclear.

I also think I still have concerns about doing anything other than lifting+shifting the exact code (ie including using the queues as-is), because that will extend the time between archiving the application-store repo and releasing the integrated assessment_store module in pre-award-stores. This would increase the time that we're blocking other teams, and/or increase the chance that we need to cherry-pick commits from the old repo to the new repo.

I think I still believe that the process should be:

1) Lift+shift exact code, release to prod and archive the old assessment-store service/repo.
2) Add the feature flag to toggle between queue+direct DB writes.
3) Switch over to direct DB writes.
4) Remove all the now-defunct queuing infrastruct.
5) Consolidate concepts (eg separate application and assessment tables - or anything of that ilk; I'm unsure of the specifics).

Would love to chat this through, as maybe I'm missing some context - or misunderstanding the proposal - and these concerns above are not as necessary to avoid as I think.

**Reply by srh-sloan** [reply]
Agree, this is a simpler approach. Have updated the description accordingly but keen to get @sfount's view on this too.
